File: app/src/main/java/com/udacity/astroapp/data/AppExecutors.java
package com.udacity.astroapp.data;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * Global executor pools for the whole application.
 */
public class AppExecutors {

    /* Singleton instantiation */
    private static final Object LOCK = new Object();
    private static AppExecutors executors;
    private final Executor diskIO;
    private final Executor mainThread;
    private final Executor networkIO;

    private AppExecutors(Executor diskIO, Executor mainThread, Executor networkIO) {
        this.diskIO = diskIO;
        this.mainThread = mainThread;
        this.networkIO = networkIO;
    }

    public static AppExecutors getExecutors() {
        if (executors == null) {
            synchronized (LOCK) {
                executors = new AppExecutors(Executors.newSingleThreadExecutor(),
                        Executors.newFixedThreadPool(3),
                        new MainThreadExecutor());
            }
        }
        return executors;
    }

    public Executor diskIO() {
        return diskIO;
    }

    @SuppressWarnings("unused")
    public Executor mainThread() {
        return mainThread;
    }

    @SuppressWarnings("unused")
    public Executor networkIO() {
        return networkIO;
    }

    private static class MainThreadExecutor implements Executor {
        private final Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(@NonNull Runnable command) {
            mainThreadHandler.post(command);
        }
    }
}

File: app/src/main/kotlin/com/udacity/astroapp/data/AppExecutors.kt
package com.udacity.astroapp.data

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * Global executor pools for the whole application.
 */
class AppExecutors private constructor(
    val diskIO: Executor,
    val mainThread: Executor,
    val networkIO: Executor
) {

    companion object {
        @Volatile
        private var executors: AppExecutors? = null

        fun getExecutors(): AppExecutors {
            return executors ?: synchronized(this) {
                executors ?: AppExecutors(
                    diskIO = Executors.newSingleThreadExecutor(),
                    networkIO = Executors.newFixedThreadPool(3),
                    mainThread = MainThreadExecutor()
                ).also { executors = it }
            }
        }
    }

    private class MainThreadExecutor : Executor {
        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }
    }
}